### PR TITLE
refactor: Remove usage of regex for portfolio URLs

### PR DESCRIPTION
**Pull request by NEllusion**
## **Description**

This PR addresses a few issues related to how we were using regex expressions. In particular, how we were using URLs in regex statements. This is due to the fact that if we use a URL in a regex statement without escaping it, the periods get interpreted as regex wildcard characters.

In this case, it means that URLs such as `https://portfolioxmetamask.io` would be be interpreted by MetaMask as being a valid portfolio URL. This could open up the door to phishing situations or unexpected behaviour.

Rather than attempting to make the regex safe to use, I opted to re-write the code without the need for regex. This uses built in Javascript URL for more accurate comparisons.

## **Related issues**

https://github.com/MetaMask/mobile-planning/issues/1571

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [ ] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [X] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [X] I’ve properly set the pull request status:
  - [X] In case it's not yet "ready for review", I've set it to "draft".
  - [X] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Review by tommasini — DISMISSED** [review]
LGTM!

**Review by NicolasMassart — DISMISSED** [review]
Awesome!

File: app/util/url/url.test.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    it('should return false for non-portfolio URLs', () => {
+      const url = 'http://www.example.com';
+      expect(isPortfolioUrl(url)).toBe(false);
+    });
+
+    it('should return false for invalid URLs', () => {
+      const url = 'invalid url';
+      expect(isPortfolioUrl(url)).toBe(false);
+    });
+  });
```
Comment (NicolasMassart) [inline app/util/url/url.test.ts:25]: Suggestion: As you provided an example of dangerous url that could pass with the regex, to confirm the issue is solved, you could add a test case matching against this `https://portfolioxmetamask.io` type of URL.

File: app/util/url/url.test.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
+import { isPortfolioUrl, isBridgeUrl } from './index';
+import AppConstants from '../../core/AppConstants';
+
+describe('URL Check Functions', () => {
+  describe('isPortfolioUrl', () => {
+    it('should return true for portfolio URLs', () => {
```
Comment (NicolasMassart) [inline app/util/url/url.test.ts:6]:
```suggestion
    it('returns true for portfolio URLs', () => {
```

File: app/util/url/url.test.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+      const url = AppConstants.PORTFOLIO.URL;
+      expect(isPortfolioUrl(url)).toBe(true);
+    });
+
+    it('should return false for URLs similar to portfolio', () => {
```
Comment (NicolasMassart) [inline app/util/url/url.test.ts:11]:
```suggestion
    it('returns false for URLs that were false positive with previous regex implementation', () => {
```

File: app/util/url/url.test.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+      const url = 'https://portfolioxmetamask.io';
+      expect(isPortfolioUrl(url)).toBe(false);
+    });
+
+    it('should return true for portfolio URLs with additional params', () => {
```
Comment (NicolasMassart) [inline app/util/url/url.test.ts:16]:
```suggestion
    it('returns true for portfolio URLs with additional params', () => {
```

File: app/util/url/url.test.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+      const url = `${AppConstants.PORTFOLIO.URL}/bridge?foo=bar`;
+      expect(isBridgeUrl(url)).toBe(true);
+    });
+
+    it('should return false for non-portfolio URLs', () => {
```
Comment (NicolasMassart) [inline app/util/url/url.test.ts:21]:
```suggestion
    it('returns false for non-portfolio URLs', () => {
```

File: app/util/url/url.test.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+      const url = 'http://www.example.com';
+      expect(isPortfolioUrl(url)).toBe(false);
+    });
+
+    it('should return false for invalid URLs', () => {
```
Comment (NicolasMassart) [inline app/util/url/url.test.ts:26]:
```suggestion
    it('returns false for invalid URLs', () => {
```

File: app/util/url/url.test.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+      const url = 'invalid url';
+      expect(isPortfolioUrl(url)).toBe(false);
+    });
+  });
+
+  describe('isBridgeUrl', () => {
+    it('should return true for bridge URLs', () => {
```
Comment (NicolasMassart) [inline app/util/url/url.test.ts:33]:
```suggestion
    it('returns true for bridge URLs', () => {
```

File: app/util/url/url.test.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+      const url = AppConstants.BRIDGE.URL;
+      expect(isBridgeUrl(url)).toBe(true);
+    });
+
+    it('should return false for URLs similar to bridge URLs', () => {
```
Comment (NicolasMassart) [inline app/util/url/url.test.ts:38]:
```suggestion
    it('returns false for URLs that were false positive with previous regex implementation', () => {
```

File: app/util/url/url.test.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+      const url = 'https://portfolioxmetamask.io/bridge';
+      expect(isPortfolioUrl(url)).toBe(false);
+    });
+
+    it('should return true for bridge URLs with additional params', () => {
```
Comment (NicolasMassart) [inline app/util/url/url.test.ts:43]:
```suggestion
    it('returns true for bridge URLs with additional params', () => {
```

File: app/util/url/url.test.ts [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+      const url = `${AppConstants.BRIDGE.URL}?foo=bar`;
+      expect(isBridgeUrl(url)).toBe(true);
+    });
+
+    it('should return true for bridge URLs with trailing slash', () => {
```
Comment (NicolasMassart) [inline app/util/url/url.test.ts:48]:
```suggestion
    it('returns true for bridge URLs with trailing slash', () => {
```

File: app/util/url/url.test.ts [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+      const url = `${AppConstants.BRIDGE.URL}/`;
+      expect(isBridgeUrl(url)).toBe(true);
+    });
+
+    it('should return false for non-bridge URLs', () => {
```
Comment (NicolasMassart) [inline app/util/url/url.test.ts:53]:
```suggestion
    it('returns false for non-bridge URLs', () => {
```

File: app/util/url/url.test.ts [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+      const url = 'http://www.example.com';
+      expect(isBridgeUrl(url)).toBe(false);
+    });
+
+    it('should return false for invalid URLs', () => {
```
Comment (NicolasMassart) [inline app/util/url/url.test.ts:58]:
```suggestion
    it('returns false for invalid URLs', () => {
```

**Review by NicolasMassart — COMMENTED** [review]
Suggestions for [test naming guidelines](https://github.com/MetaMask/contributor-docs/blob/main/docs/unit-testing.md#use-it-to-specify-the-desired-behavior-for-the-code-under-test) compliance.
Other than that, it's good to me.
Make sure to check if you need review from another team before merge!

**Review by sethkfman — DISMISSED** [review]
LGTM

**Review by sethkfman — APPROVED** [review]
LGTM

**Pull request merged**